File: 729. My Calendar I/solution.js
var MyCalendar = function() {
    this.calendar = [];
};

/** 
 * @param {number} start 
 * @param {number} end
 * @return {boolean}
 */
 
MyCalendar.prototype.book = function(start, end) {
    for (let [s, e] of this.calendar) {
        if (!(end <= s || start >= e)) { 
            return false;
        }
    }
    this.calendar.push([start, end]);
    return true;
};

File: 729. My Calendar I/solution.java
class MyCalendar {

    private TreeSet<int[]> calendar;

    public MyCalendar() {
        calendar = new TreeSet<>((a, b) -> a[0] - b[0]);
    }
    
    public boolean book(int start, int end) {
        int[] newEvent = {start, end};
        int[] prev = calendar.floor(newEvent);
        int[] next = calendar.ceiling(newEvent);

        if (prev != null && prev[1] > start) {
            return false; 
        }
        if (next != null && next[0] < end) {
            return false; 
        }

        calendar.add(newEvent);
        return true;
    }
}